external-crates/move/crates/move-compiler/src/naming/translate.rs

```diff
@@ -406,89 +492,99 @@ impl<'env> Context<'env> {
         resolved
     }
 
-    fn resolve_module_type(&mut self, loc: Loc, m: &ModuleIdent, n: &Name) -> Option<ModuleType> {
-        let types = match self.scoped_types.get(m) {
-            None => {
-                self.env.add_diag(diag!(
-                    NameResolution::UnboundModule,
-                    (m.loc, format!("Unbound module '{}'", m)),
-                ));
-                return None;
-            }
-            Some(members) => members,
-        };
-        match types.get(&n.value) {
-            None => {
-                let msg = format!(
-                    "Invalid module access. Unbound struct '{}' in module '{}'",
-                    n, m
-                );
-                self.env
-                    .add_diag(diag!(NameResolution::UnboundModuleMember, (loc, msg)));
-                None
-            }
-            Some(module_type) => Some(module_type.clone()),
+    fn resolve_module_type_opt(&self, m: &ModuleIdent, n: &Name) -> Option<ModuleType> {
+        if let Some(types) = self.scoped_types.get(m) {
+            types.get(&n.value).cloned()
+        } else {
+            None
         }
     }
 
-    fn resolve_module_function(
+    fn resolve_function_opt(&self, m: &ModuleIdent, n: &Name) -> Option<FunctionName> {
+        self.scoped_functions
+            .get(m)
+            .and_then(|functions| functions.get(&n.value).map(|_| FunctionName(*n)))
+    }
+
+    fn resolve_constant_opt(&self, m: &ModuleIdent, n: &Name) -> Option<ConstantName> {
+        self.scoped_constants
+            .get(m)
+            .and_then(|constants| constants.get(&n.value).map(|_| ConstantName(*n)))
+    }
+
+    fn resolve_module_access(
         &mut self,
-        loc: Loc,
         m: &ModuleIdent,
         n: &Name,
-    ) -> Option<FunctionName> {
-        let functions = match self.scoped_functions.get(m) {
-            None => {
-                self.env.add_diag(diag!(
-                    NameResolution::UnboundModule,
-                    (m.loc, format!("Unbound module '{}'", m)),
-                ));
-                return None;
+        expected_access_type: ModuleAccessKind,
+    ) -> Option<ResolvedModuleAccess> {
+        let function_opt = self
+            .resolve_function_opt(m, n)
+            .map(ResolvedModuleAccess::Function);
+        let constant_opt = self
+            .resolve_constant_opt(m, n)
+            .map(ResolvedModuleAccess::Constant);
+        let datatype_opt = self
+            .resolve_module_type_opt(m, n)
+            .map(ResolvedModuleAccess::Datatype);
+        match expected_access_type {
```

Comment:
What do you think about something like this to remove some of the nesting/if-else statements?

```rust
ModuleAccess::Function => {
    function_opt.or(constant_opt).or(datatype_opt)
}
ModuleAccess::Constant => {
    constant_opt.or(datatype_opt).or(function_opt)
}
ModuleAccess::Datatype => {
    datatype_opt.or(constant_opt).or(function_opt)
}

Reply:
I would have done that if I knew `or` existed -- I even looked at `std::option` looking for it